packages/material-ui-utils/src/elementTypeAcceptingRef.js

```diff
@@ -0,0 +1,50 @@
+import * as PropTypes from 'prop-types';
+import React from 'react';
+import { isLazy, isMemo } from 'react-is';
+import chainPropTypes from './chainPropTypes';
+
+function isClassComponent(elementType) {
+  // elementType.prototype?.isReactComponent
+  const { prototype = {} } = elementType;
+
+  return Boolean(prototype.isReactComponent);
+}
+
+function elementTypeAcceptingRef(props, propName, componentName, location, propFullName) {
+  const propValue = props[propName];
+  const safePropName = propFullName || propName;
+
+  if (propValue == null) {
+    return null;
+  }
+
+  let warningHint;
+
+  /**
+   * blacklisting instead of whitelisting
+   *
+   * blacklisting will miss some components like React.Fragment. Those will at least
+   * trigger a warning in react.
+   * We can't whitelist because there is no safe way to detect React.forwardRef
+   * or class components. Safe means there's no public API.
+   *
+   */
+  if (isLazy(propValue)) {
+    warningHint = 'But you passed a React.lazy component.';
+  } else if (isMemo(propValue)) {
+    warningHint = 'But you passed a React.memo component.';
+  } else if (typeof propValue === 'function' && !isClassComponent(propValue)) {
+    warningHint = 'Did you accidentally provide a plain function component instead?';
+  }
+
+  if (warningHint !== undefined) {
+    return new Error(
+      `Invalid ${location} \`${safePropName}\` supplied to \`${componentName}\`. ` +
+        `Expected an element type that can hold a ref. ${warningHint}`,
```

Comment:
Should we link to the Caveat with refs section of the composition guide here?

Reply:
> add composition docs link to prop types warning

Yes. The goal is that you shouldn't have to google for a solution. It should be obvious what the problem and solution is simply by following the instructions given by the error.

Reply:
I don't know if we're talking at cross purposes, but I was making a suggestion, while you seem to be saying this is already the case, which it doesn't appear to be from the code.

>> add composition docs link to prop types warning

Not sure what's being quoted from here? It looks like a commit description, but I don't see that commit here...

Reply:
It's from the PR description under "todo". I was being snarky.

Reply:
I see.